internal.nix

```diff
@@ -29,6 +29,26 @@ rec {
   makeValidDrvName = str:
     lib.stringAsChars (c: if isValidDrvNameChar c then c else "?") str;
 
+  # Description: Checks if a string looks like a valid git revision
+  # Type: String -> Boolean
+  isGitRev = str:
+    (builtins.match "[0-9a-f]{40}" str) != null;
+
+  # Description: Takes a string of the format "git+http(s)://domain.tld/repo#branch" and returns
+  # an attribute set { url, rev }
+  # Type: String -> Set
+  parseGitRef = str:
+    let
+      parts = builtins.split "[#]" str;
+    in
+    assert !(builtins.length parts == 3) ->
+      builtins.throw "[npmlock2nix] failed to parse Git reference `${str}`. Expected a string of format `git+http(s)://domain.tld/repo#branch`";
```

Comment:
The branch is optional so `builtins.length parts == 1` should also be allowed. This should be fine since fetchGit's `ref` argument is optional.

Reply:
https://github.com/Sohalt/npmlock2nix/pull/1

Reply:
Please drop the `[npmlock2nix]` prefix in the message and use the `throw` function (which is in scope and defined at the top of this file).

Reply:
@ryneeverett what branch / commit would this default to if the length is 1? What revision should we fetch in that case? It feels a bit ambigious. This probably works for the `package.json` entries but wouldn't work for the actual lockfiles (or at least wouldn't really be reproducible).

Reply:
> what branch / commit would this default to if the length is 1?

I thought I'd read somewhere that it pointed to the default branch (`HEAD`) but according to these docs it's `master`. https://docs.npmjs.com/cli/v7/configuring-npm/package-json#git-urls-as-dependencies

> It feels a bit ambigious. This probably works for the package.json entries but wouldn't work for the actual lockfiles (or at least wouldn't really be reproducible).

Correct, lockfiles aren't actually reproducible as these git dependencies get passed through to the lockfile while still underspecified. I suppose packages which specify ambiguous dependencies just need to be rejected in order to support npmlock2nix's goal of supporting IFD?

Reply:
> Correct, lockfiles aren't actually reproducible as these git dependencies get passed through to the lockfile while still underspecified.
I don't really care about the lockfiles being reproducible in this case. As long as we have a lockfile we should be able to reproduce whatever it currently expresses. How `npm` manages to reproduce the files is fortunately not my "job" :-)

>I suppose packages which specify ambiguous dependencies just need to be rejected in order to support npmlock2nix's goal of supporting IFD?

Where do we state that IFD is a goal? I am not sure I can follow that argument. Usually you can use IFD with anything that wasn't made for IFD. I wouldn't add a requirement or special feature for IFD.

Reply:
> Where do we state that IFD is a goal?

That was my, perhaps erroneous, interpretation of "Works in restricted evaluation" which I took to mean npmlock2nix can be used in a purely evaluated nix expression. Seems like this would be incompatible with non-deterministically fetching the latest revision.

I don't have a strong grasp on all these terms and concepts and I could be wildly confused. :)

Reply:
I'm going to accept that it appears that lockfiles always have a commit in the version even if it's not specified in the package.json, so this point appears moot.